### PR TITLE
Add SST ingestion to ldb

### PR DESCRIPTION
**Pull request by riversand963**
We add two subcommands `write_extern_sst` and `ingest_extern_sst` to ldb. This PR avoids changing existing code because we hope to cherry-pick to earlier releases to support compatibility check for external SST file ingestion.

Test plan:
```
$make clean && make -j16 ldb
$./ldb --db=<path/to/db> write_extern_sst <path/to/extern/sst>
$./ldb --db=<path/to/db> ingest_extern_sst <path/to/input/sst> [--allow_global_seqno] [...]
$python tools/ldb_test.py
```

### CODE REVIEW

File: tools/ldb_cmd.cc [hunk 1]
```diff
@@ -2939,5 +2947,166 @@ void DBFileDumperCommand::DoCommand() {
...
+  if (params.size() != 1) {
+    exec_state_ = LDBCommandExecuteResult::Failed(
+        "output SST file path must be specified");
+  } else {
+    output_sst_path_ = params.at(0);
+  }
+}
+
+void WriteExternalSstFilesCommand::DoCommand() {
+  if (!db_) {
```
Comment (siying) [inline tools/ldb_cmd.cc]: Hmm, we got a DB to write external SST files? It feels a little bit odd to me.

File: tools/ldb_cmd.cc [hunk 2]
```diff
@@ -2939,5 +2947,166 @@ void DBFileDumperCommand::DoCommand() {
...
+  }
+
+  status = sst_file_writer.Finish();
+  if (!status.ok()) {
+    return;
+  }
+
+  if (bad_lines > 0) {
+    fprintf(stderr, "Warning: %d bad lines ignored.\n", bad_lines);
+  }
```
Comment (siying) [inline tools/ldb_cmd.cc]: I didn't see where you set `exec_state_`.  Will it still return non-0 return code?

File: tools/ldb_cmd.cc [hunk 3]
```diff
@@ -2939,5 +2947,166 @@ void DBFileDumperCommand::DoCommand() {
...
+  ifo.move_files = move_files_;
+  ifo.snapshot_consistency = snapshot_consistency_;
+  ifo.allow_global_seqno = allow_global_seqno_;
+  ifo.allow_blocking_flush = allow_blocking_flush_;
+  ifo.ingest_behind = ingest_behind_;
+  ifo.write_global_seqno = write_global_seqno_;
+  Status status = db_->IngestExternalFile(cfh, {input_sst_path_}, ifo);
+  if (!status.ok()) {
+    fprintf(stderr, "Failed to ingest external SSTs: %s\n",
+            status.ToString().c_str());
```
Comment (siying) [inline tools/ldb_cmd.cc]: Same question. I didn't see exec_state_ is set. Does the failure runs return non-0 return status?

File: tools/ldb_cmd.cc (hunk 1 above)
Comment (riversand963) [reply tools/ldb_cmd.cc]: I think we need a `ColumnFamilyHandle` to write external SST files.

Comment (siying) [reply tools/ldb_cmd.cc]: I think it's OK to pass nullptr.

Comment (siying) [reply tools/ldb_cmd.cc]: But anyway it's a good idea to test non-nullptr case.

File: tools/ldb_cmd.cc (hunk 2 above)
Comment (riversand963) [reply tools/ldb_cmd.cc]: Done.

File: tools/ldb_cmd.cc (hunk 3 above)
Comment (riversand963) [reply tools/ldb_cmd.cc]: Done.

File: tools/ldb_cmd.cc (hunk 1 above)
Comment (riversand963) [reply tools/ldb_cmd.cc]: Another benefit is that we can have the flexibility to iterate over an open db and write all its records to an external SST.

**Comment by riversand963** [comment]
@siying I've added a unit test in `ldb_test.py`. Different versions of RocksDB support different ingestion options. Therefore, the test added in this PR is a minimum.

File: tools/ldb_test.py [hunk 4]
```diff
@@ -547,5 +554,25 @@ def testColumnFamilies(self):
         # non-existing column family.
         self.assertRunFAIL("get cf3_1 --column_family=four")
 
+    def testIngestExternalSst(self):
+        print "Running testIngestExternalSst..."
+
+        # Dump, load, write external sst and ingest it in another db
+        dbPath = os.path.join(self.TMP_DIR, "db1")
+        self.assertRunOK("batchput --db=%s --create_if_missing x1 y1 x2 y2 x3 y3 x4 y4" % dbPath, "OK")
```
Comment (siying) [inline tools/ldb_test.py]: 80-char

Comment (riversand963) [reply tools/ldb_test.py]: Looks like `make format` does not check python files. I'll manually fix.

**Comment by ajkr** [comment]
Nice! I've always used `ldb repair` to create a DB out of random SST files I have lying around but will now use this feature instead :).

**Comment by riversand963** [comment]
@ajkr My plan is to backport this feature to earlier versions of RocksDB. Doing so may require resolving conflicts or code updates. Once it's done, we can add external SST file ingestion to compatibility test.

**Comment by ytrezq** [comment]
@riversand963 it seems the `write_extern_sst` command creates an empty sst file and stalls in the case of multiples column families databases (so no longer works).

**Pull request closed without merging**